### PR TITLE
[Choreo] Changes to consider organization with subscription policies

### PR DESCRIPTION
**Pull request by slahirucd7**
### Purpose
$subject
Relevant issue: https://github.com/wso2-enterprise/choreo/issues/27601

### Issues

Fixes #

### Automation tests
 - Unit tests added: Yes
 - Integration tests added: No

### Tested environments

Not Tested

---
#### Maintainers: Check before merge
- [x] Assigned 'Type' label
- [x] Assigned the project
- [x] Validated respective github issues
- [x] Assigned milestone to the github issue(s)

### CODE REVIEW

File: adapter/internal/discovery/xds/rate_limiter_cache_test.go [hunk 1]
```diff
@@ -577,3 +578,84 @@ func getDummyAPISwagger(apiID, level, apiPolicy, res1GetPolicy, res1PostPolicy,
...
+						RequestCount: 300,
+						TimeUnit:     "HOUR",
+					},
+				},
+			},
+		},
+	}
+
+	// Initialize rlsPolicyCache.metadataBasedPolicies
+	rlsPolicyCache.metadataBasedPolicies = make(map[string]map[string][]*rls_config.RateLimitDescriptor)
```
Comment (slahirucd7) [inline adapter/internal/discovery/xds/rate_limiter_cache_test.go:628]: Consider org as well.

File: adapter/internal/discovery/xds/rate_limiter_cache.go [hunk 2]
```diff
@@ -77,7 +80,7 @@ type rateLimitPolicyCache struct {
 	apiLevelRateLimitPolicies map[string]map[string]map[string][]*rls_config.RateLimitDescriptor
 	// metadataBasedPolicies is used to store the rate limit policies which are based on dynamic metadata.
 	// rate limit type (eg: subscription) -> policy name (eg: Gold, Silver) -> Rate Limit Config
```
Comment (renuka-fernando) [inline adapter/internal/discovery/xds/rate_limiter_cache.go:82]: Let's update this comment as well. Include the Org in the correct place.

File: adapter/internal/discovery/xds/rate_limiter_cache.go [hunk 3]
```diff
@@ -247,16 +250,26 @@ func (r *rateLimitPolicyCache) generateRateLimitConfig(label string) *rls_config
...
 	//Iterate through the subscription policies and append it to the orgDescriptors
-	for metadataType, metadataPolicyMap := range r.metadataBasedPolicies {
-		var subscriptionDescriptors []*rls_config.RateLimitDescriptor
-		for _, subscriptionPolicyDescriptor := range metadataPolicyMap {
-			subscriptionDescriptors = append(subscriptionDescriptors, subscriptionPolicyDescriptor)
-		}
-		metadataDescriptor := &rls_config.RateLimitDescriptor{
-			Key:         metadataType,
-			Descriptors: subscriptionDescriptors,
+	//organization -> subscription -> policyName -> RateLimitDescriptor
```
Comment (renuka-fernando) [inline adapter/internal/discovery/xds/rate_limiter_cache.go:253]:
```suggestion
	// subscription -> organization -> policyName -> RateLimitDescriptor
```

Comment (renuka-fernando) [inline adapter/internal/discovery/xds/rate_limiter_cache.go:253]: Shall we update this comment with a simple YAML definition of a Rate-Limiter? It is more descriptive than this comment.

File: adapter/internal/discovery/xds/rate_limiter_cache.go [hunk 4]
```diff
@@ -247,16 +250,26 @@ func (r *rateLimitPolicyCache) generateRateLimitConfig(label string) *rls_config
...
+	if subscriptionPoliciesList, ok := r.metadataBasedPolicies[subscriptionPolicyType]; ok {
+		for orgName := range subscriptionPoliciesList {
```
Comment (renuka-fernando) [inline adapter/internal/discovery/xds/rate_limiter_cache.go:255]:
```suggestion
		for orgUUID := range subscriptionPoliciesList {
```

File: adapter/internal/discovery/xds/rate_limiter_cache_test.go [hunk 5]
```diff
@@ -577,3 +578,86 @@ func getDummyAPISwagger(apiID, level, apiPolicy, res1GetPolicy, res1PostPolicy,
...
+						TimeUnit:     "hours",
+					},
+				},
+				Organization: "org1",
+			},
+		},
+	}
+
+	// Initialize rlsPolicyCache.metadataBasedPolicies
+	rlsPolicyCache.metadataBasedPolicies = make(map[string]map[string]map[string]*rls_config.RateLimitDescriptor)
```
Comment (renuka-fernando) [inline adapter/internal/discovery/xds/rate_limiter_cache_test.go:632]: This is initialized inside `AddSubscriptionLevelRateLimitPolicy` neh? so we can remove this line

File: adapter/internal/discovery/xds/rate_limiter_cache_test.go [hunk 6]
```diff
@@ -577,3 +578,86 @@ func getDummyAPISwagger(apiID, level, apiPolicy, res1GetPolicy, res1PostPolicy,
...
+	expectedPolicies := map[string]map[string]map[string]*rls_config.RateLimitDescriptor{
+		subscriptionPolicyType: {
+			"org1": {
+				"Policy1": {
+					Key:   "policy",
+					Value: "Policy1",
+					RateLimit: &rls_config.RateLimitPolicy{
+						Unit:            rls_config.RateLimitUnit_SECOND,
+						RequestsPerUnit: 100,
+						Name:            "Policy1",
```
Comment (renuka-fernando) [inline adapter/internal/discovery/xds/rate_limiter_cache_test.go:646]: Let's remove this `Name`

File: adapter/internal/discovery/xds/rate_limiter_cache_test.go [hunk 7]
```diff
@@ -577,3 +578,86 @@ func getDummyAPISwagger(apiID, level, apiPolicy, res1GetPolicy, res1PostPolicy,
...
+				},
+				"Policy2": {
+					Key:   "policy",
+					Value: "Policy2",
+					RateLimit: &rls_config.RateLimitPolicy{
+						Unit:            rls_config.RateLimitUnit_MINUTE,
+						RequestsPerUnit: 200,
+						Name:            "Policy2",
+					},
+				},
```
Comment (renuka-fernando) [inline adapter/internal/discovery/xds/rate_limiter_cache_test.go:657]: Why `Unauthenticated` policy is not there in the list?

File: adapter/internal/discovery/xds/rate_limiter_cache_test.go [hunk 8]
```diff
@@ -577,3 +578,86 @@ func getDummyAPISwagger(apiID, level, apiPolicy, res1GetPolicy, res1PostPolicy,
 	mgwSwagger.VHost = "vhost1"
 	return mgwSwagger
 }
+func TestAddSubscriptionLevelRateLimitPolicy(t *testing.T) {
+	policyList := &types.SubscriptionPolicyList{
```
Comment (renuka-fernando) [inline adapter/internal/discovery/xds/rate_limiter_cache_test.go:582]: Shall we add a second event too with a second org `org2`?

File: adapter/internal/discovery/xds/rate_limiter_cache.go [hunk 9]
```diff
@@ -317,7 +331,10 @@ func AddSubscriptionLevelRateLimitPolicy(policyList *types.SubscriptionPolicyLis
 		}
 
 		// Need not to add the Unauthenticated and Unlimited policies to the rate limiter service
-		if policy.Name == "Unauthenticated" || policy.Name == "Unlimited" {
+		if policy.Organization == "carbon.super" && policy.Name == "Unauthenticated" {
+			continue
+		}
+		if policy.Name == "Unlimited" {
```
Comment (renuka-fernando) [inline adapter/internal/discovery/xds/rate_limiter_cache.go:337]: Without using the policy name `Unlimited` to check, shall we use RequestsPerUnit as `-1`?
@malinthaprasan, if a user deletes the default `Unlimited` policy and creates a new one with a different name for the Unlimited policy, do we have to handle this Unlimited case neh?

When creating a policy from admin API, should we have the option to mark a policy as unlimited? or maybe set RequestsPerUnit as -1?

File: adapter/internal/discovery/xds/rate_limiter_cache_test.go [hunk 10]
```diff
@@ -608,8 +608,8 @@ func TestAddSubscriptionLevelRateLimitPolicy(t *testing.T) {
 				DefaultLimit: &types.SubscriptionDefaultLimit{
 					QuotaType: "requestCount",
 					RequestCount: &types.SubscriptionRequestCount{
-						RequestCount: 300,
-						TimeUnit:     "hours",
+						RequestCount: 2147483647,
+						TimeUnit:     "min",
```
Comment (renuka-fernando) [inline adapter/internal/discovery/xds/rate_limiter_cache_test.go:612]: Keep these as it is

File: adapter/internal/discovery/xds/rate_limiter_cache_test.go [hunk 11]
```diff
@@ -608,8 +608,8 @@ func TestAddSubscriptionLevelRateLimitPolicy(t *testing.T) {
 				DefaultLimit: &types.SubscriptionDefaultLimit{
```
Comment (renuka-fernando) [inline adapter/internal/discovery/xds/rate_limiter_cache_test.go:608]: Let's add Unlimited for org1 and carbon.super

Comment (slahirucd7) [reply adapter/internal/discovery/xds/rate_limiter_cache_test.go:608]: Added with: [e5ebf4f](https://github.com/wso2/product-microgateway/pull/3504/commits/e5ebf4fd5bc9252ff351408c2303b9df29314601)

File: adapter/internal/discovery/xds/rate_limiter_cache.go [hunk 12]
```diff
@@ -76,8 +79,8 @@ type rateLimitPolicyCache struct {
 	// org -> vhost -> API-Identifier (i.e. Vhost:API-UUID) -> Rate Limit Configs
 	apiLevelRateLimitPolicies map[string]map[string]map[string][]*rls_config.RateLimitDescriptor
 	// metadataBasedPolicies is used to store the rate limit policies which are based on dynamic metadata.
-	// rate limit type (eg: subscription) -> policy name (eg: Gold, Silver) -> Rate Limit Config
-	metadataBasedPolicies map[string]map[string]*rls_config.RateLimitDescriptor
+	// metadata related to the subscription rate-limiting: organization -> subscriptionID -> rate-limit-policy
```
Comment (VirajSalaka) [inline adapter/internal/discovery/xds/rate_limiter_cache.go:82]: Is this comment wrong ?
rate limit Type (ex: subscription) -> organization -> policyName -> policy

Comment (renuka-fernando) [reply adapter/internal/discovery/xds/rate_limiter_cache.go:82]: Yeah it should be as @VirajSalaka mentioned

Comment (slahirucd7) [reply adapter/internal/discovery/xds/rate_limiter_cache.go:82]:
```
domain: Default
descriptors:
  - key: organisation
     Value: org001
        - key: subscription
     descriptors:
      - key: policy
        value: gold
        rate_limit:
          requests_per_unit: 1000
          unit: minute
      - key: policy
        value: silver
        rate_limit:
          requests_per_unit: 200
          unit: minute
```

This is how we agreed to arranging descriptors right ?

File: adapter/internal/discovery/xds/rate_limiter_cache.go [hunk 13]
```diff
@@ -319,13 +328,22 @@ func AddSubscriptionLevelRateLimitPolicy(policyList *types.SubscriptionPolicyLis
...
+		descriptor := &rls_config.RateLimitDescriptor{
 			Key:       "policy",
 			Value:     policy.Name,
 			RateLimit: &rlPolicyConfig,
 		}
-		loggers.LoggerXds.Infof("Subscription level rate limit policy is added to the cache map: %v", policy)
+		// Check if rlsPolicyCache.metadataBasedPolicies[Subscription] exists and create a new map if not
+		if _, ok := rlsPolicyCache.metadataBasedPolicies[subscriptionPolicyType]; !ok {
+			rlsPolicyCache.metadataBasedPolicies[subscriptionPolicyType] = make(map[string]map[string]*rls_config.RateLimitDescriptor)
+		}
```
[the remark this answers is not in the document]
Comment (slahirucd7) [reply adapter/internal/discovery/xds/rate_limiter_cache.go:357]: As we discussed this is required for the unit test

File: adapter/internal/discovery/xds/rate_limiter_cache.go [hunk 14]
```diff
@@ -331,10 +331,7 @@ func AddSubscriptionLevelRateLimitPolicy(policyList *types.SubscriptionPolicyLis
 		}
 
 		// Need not to add the Unauthenticated and Unlimited policies to the rate limiter service
-		if policy.Organization == "carbon.super" && policy.Name == "Unauthenticated" {
-			continue
-		}
-		if policy.Name == "Unlimited" {
+		if (policy.Organization == "carbon.super" && policy.Name == "Unauthenticated") || policy.DefaultLimit.RequestCount.RequestCount == -1 {
```
Comment (renuka-fernando) [inline adapter/internal/discovery/xds/rate_limiter_cache.go:334]:
```suggestion
		if (policy.Organization == "carbon.super" && policy.Name == "Unauthenticated") || policy.DefaultLimit.RequestCount.RequestCount < 0 {
```

File: adapter/internal/discovery/xds/rate_limiter_cache.go (hunk 12 above)
Comment (renuka-fernando) [reply adapter/internal/discovery/xds/rate_limiter_cache.go:82]: yes, but the internal map `metadataBasedPolicies` contains in the following format

```yaml
subscription:
  org1:
    gold: {descriptorObject: {}}
    silver: {descriptorObject: {}}
  org2:
    gold: {descriptorObject: {}}
```

File: adapter/internal/discovery/xds/rate_limiter_cache.go (hunk 13 above)
[the remark this answers is not in the document]
Comment (renuka-fernando) [reply adapter/internal/discovery/xds/rate_limiter_cache.go:354]: Let move this to original place

File: enforcer-parent/enforcer/src/main/java/org/wso2/choreo/connect/enforcer/security/jwt/JWTAuthenticator.java [hunk 15]
```diff
@@ -314,10 +319,35 @@ public AuthenticationContext authenticate(RequestContext requestContext) throws
...
+                        }
+                        SubscriptionDataStore datastore = SubscriptionDataHolder.getInstance()
+                                .getTenantSubscriptionStore(apiTenantDomain);
                         String subscriptionId = authenticationContext.getApiUUID() + ":" +
                                 authenticationContext.getApplicationUUID();
+                        String subPolicyName = authenticationContext.getTier();
                         requestContext.addMetadataToMap("ratelimit:subscription", subscriptionId);
-                        requestContext.addMetadataToMap("ratelimit:usage-policy", authenticationContext.getTier());
+                        requestContext.addMetadataToMap("ratelimit:usage-policy", subPolicyName);
+                        String orgList = System.getenv("CUSTOM_SUBSCRIPTION_POLICY_HANDLING_ORG");
```
Comment (renuka-fernando) [inline enforcer-parent/enforcer/src/main/java/org/wso2/choreo/connect/enforcer/security/jwt/JWTAuthenticator.java:334]: Shall we read this once, maybe as a static var? We can keep this feature flag in the top part of the file.

File: enforcer-parent/enforcer/src/main/java/org/wso2/choreo/connect/enforcer/security/jwt/JWTAuthenticator.java [hunk 16]
```diff
@@ -314,10 +319,35 @@ public AuthenticationContext authenticate(RequestContext requestContext) throws
...
+                        if (datastore.getSubscriptionPolicyByName(subPolicyName) != null &&
+                                StringUtils.isNotEmpty(orgList)) {
+                            SubscriptionPolicy subPolicy = datastore.getSubscriptionPolicyByName(subPolicyName);
+                            Set<String> orgSet = Stream.of(orgList.trim().split("\\s*,\\s*"))
+                                    .collect(Collectors.toSet());
```
Comment (renuka-fernando) [inline enforcer-parent/enforcer/src/main/java/org/wso2/choreo/connect/enforcer/security/jwt/JWTAuthenticator.java:339]: We can keep this in a static var, we can evaluate this only once.

File: enforcer-parent/enforcer/src/main/java/org/wso2/choreo/connect/enforcer/security/jwt/JWTAuthenticator.java [hunk 17]
```diff
@@ -314,10 +319,35 @@ public AuthenticationContext authenticate(RequestContext requestContext) throws
...
+                            if (StringUtils.isNotEmpty(subPolicy.getOrganization()) &&
+                                    orgSet.contains(subPolicy.getOrganization()) || orgSet.contains("*")) {
```
Comment (renuka-fernando) [inline enforcer-parent/enforcer/src/main/java/org/wso2/choreo/connect/enforcer/security/jwt/JWTAuthenticator.java:341]:
```suggestion
                            if (StringUtils.isNotEmpty(subPolicy.getOrganization()) &&
                                    (orgSet.contains(subPolicy.getOrganization()) || orgList == "*"))) {
```

File: enforcer-parent/enforcer/src/main/java/org/wso2/choreo/connect/enforcer/security/jwt/JWTAuthenticator.java (hunk 15 above)
Comment (slahirucd7) [reply enforcer-parent/enforcer/src/main/java/org/wso2/choreo/connect/enforcer/security/jwt/JWTAuthenticator.java:334]: We don't need to maintain this if we are using static vars.

File: enforcer-parent/enforcer/src/main/java/org/wso2/choreo/connect/enforcer/security/jwt/JWTAuthenticator.java (hunk 17 above)
Comment (slahirucd7) [reply enforcer-parent/enforcer/src/main/java/org/wso2/choreo/connect/enforcer/security/jwt/JWTAuthenticator.java:341]: Better to use equals.

**Review by renuka-fernando — DISMISSED** [review]

File: adapter/internal/discovery/xds/rate_limiter_cache_test.go [hunk 18]
```diff
@@ -577,3 +578,135 @@ func getDummyAPISwagger(apiID, level, apiPolicy, res1GetPolicy, res1PostPolicy,
...
+				},
+				Organization: "org2",
+			},
+		},
+	}
+
+	// Initialize rlsPolicyCache.metadataBasedPolicies
+	rlsPolicyCache.metadataBasedPolicies = make(map[string]map[string]map[string]*rls_config.RateLimitDescriptor)
+
+	err := AddSubscriptionLevelRateLimitPolicy(policyList)
```
Comment (VirajSalaka) [inline adapter/internal/discovery/xds/rate_limiter_cache_test.go:667]: Shall we do an assertion on what is stored and what is not too.

**Review by renuka-fernando — APPROVED** [review]

**Review by renuka-fernando — APPROVED** [review]

**Comment by choreo-cicd** [comment]
[succeeded] Controlplane cluster : dev-deployment-v2 : 20240508.25

**Comment by choreo-cicd** [comment]
[succeeded] Dataplane(EastUS) cluster : dev-deployment-v2 : 20240508.25

**Comment by choreo-cicd** [comment]
[succeeded] Dataplane(NorthEU) cluster : dev-deployment-v2 : 20240508.25

**Comment by choreo-cicd** [comment]
[succeeded]  : dev-deployment-v2 : 20240508.25

**Comment by choreo-cicd** [comment]
[failed] Controlplane cluster : stage-deployment-v2 : 20240513.3

**Comment by choreo-cicd** [comment]
[failed] Dataplane(NorthEU) cluster : stage-deployment-v2 : 20240513.4

**Comment by choreo-cicd** [comment]
[failed] Dataplane(CDP2-WU2) cluster : stage-deployment-v2 : 20240513.4

**Comment by choreo-cicd** [comment]
[failed] Dataplane(EastUS) cluster : stage-deployment-v2 : 20240513.4

**Comment by choreo-cicd** [comment]
[failed] Dataplane(CDP2-WU2) cluster : stage-deployment-v2 : 20240513.5

**Comment by choreo-cicd** [comment]
[failed] Dataplane(EastUS) cluster : stage-deployment-v2 : 20240513.5

**Comment by choreo-cicd** [comment]
[] Dataplane(NorthEU) cluster : stage-deployment-v2 : 20240513.6

**Comment by choreo-cicd** [comment]
[] Controlplane cluster : stage-deployment-v2 : 20240513.6

**Comment by choreo-cicd** [comment]
[] Dataplane(CDP2-WU2) cluster : prod-deployment-v2 : 20240515.2

**Comment by choreo-cicd** [comment]
[] Dataplane(EastUS) cluster : prod-deployment-v2 : 20240515.2

**Comment by choreo-cicd** [comment]
[] Dataplane(EastUS) cluster : prod-deployment-v2 : 20240515.2

**Comment by choreo-cicd** [comment]
[] Controlplane cluster : prod-deployment-v2 : 20240515.2

**Comment by choreo-cicd** [comment]
[] Dataplane(NorthEU) cluster : prod-deployment-v2 : 20240515.2

**Comment by choreo-cicd** [comment]
[] Dataplane(NorthEU) cluster : prod-deployment-v2 : 20240515.2

**Pull request merged**